tests/test_layabout.py

```diff
@@ -1 +1,327 @@
-import layabout  # noqa: F401
+import os
+from unittest.mock import MagicMock, call
+
+import pytest
+
+from layabout import (
+    Layabout,
+    MissingSlackToken,
+    FailedConnection,
+)
+
+# The likelihood of this ever being a valid Slack token is probably slim.
+TOKEN = 'xoxb-13376661337-DeAdb33Fd15EA53fACef33D1'
+
+
+@pytest.fixture
+def slack_client():
+    def api_call(method):
+        return dict(
+            channels=[{'id': 'D3ADB33F1', 'name': 'general'}],
+            groups=[{'id': '1800IDGAF', 'name': 'super_secret_club'}],
+        )
+
+    slack_client = MagicMock(api_call=api_call)
+    return slack_client
+
+
+def test_layabout_can_register_handlers():
+    """
+    Test that layabout can register Slack API event handlers normally and as a
+    decorator, both with and without keyword arguments.
+    """
+    layabout = Layabout(name='test')
+    kwargs = dict(spam='🍳')
+
+    def handle_hello1(slack, event):
+        pass
+
+    @layabout.handle('hello')
+    def handle_hello2(slack, event):
+        pass
+
+    def handle_hello3(slack, event, spam):
+        pass
+
+    @layabout.handle('hello', kwargs=kwargs)
+    def handle_hello4(slack, event, spam):
+        pass
+
+    layabout.handle(type='hello')(fn=handle_hello1)
+    layabout.handle(type='hello', kwargs=kwargs)(fn=handle_hello3)
+
+
+def test_layabout_raises_type_error_with_invalid_handlers():
+    """
+    Test that layabout raises a TypeError if an event handler is supplied that
+    doesn't meet the minimum criteria to be called correctly.
+    """
+    layabout = Layabout(name='test')
+
+    def invalid_handler1(slack):
```

Comment:
The naming of function with `n` appended to end seems odd to me. If a function is different enough to get a new `n` at the end, is it not different enough to get a meaningful name variation?

Reply:
Not really, IMHO. They're both invalid handlers and they're really invalid in the same way. If anything I'd be inclined to name them the same thing.
```python
def insufficient_args(slack):
    pass

with pytest.raises(TypeError) as func_exc:
    layabout.handle(type='hello')(fn=insufficient_args)

with pytest.raises(TypeError) as decorator_exc:
    @layabout.handle('hello')
    def insufficient_args(slack):
        pass
```
I could also just shorten the first one to
```python
with pytest.raises(TypeError) as func_exc:
    layabout.handle(type='hello')(fn=lambda s: ...)
```